src/workerd/jsg/modules-new.c++

```diff
@@ -0,0 +1,1401 @@
+  bundles_[kFallback] = builder->bundles_[kFallback].releaseAsArray();
+}
+
+void ModuleRegistry::attachToIsolate(
+    Lock& js,
+    const CompilationObserver& observer) {
+  // We use new here because it is required for the v8 integration.
+  // Responsibility for cleaning up the type is passed to v8 when
+  // the context is cleaned up.
+  new IsolateModuleRegistry(js, *this, observer);
+}
+
+kj::Maybe<Module&> ModuleRegistry::resolve(const ResolveContext& context) {
+  constexpr auto tryFind = [](
+      const ResolveContext& context,
+      kj::ArrayPtr<kj::Own<ModuleBundle>> bundles) -> kj::Maybe<Module&> {
+    for (auto& bundle : bundles) {
+      KJ_IF_SOME(found, bundle->resolve(context)) {
+        return found;
+      }
+    }
+    return kj::none;
+  };
+
+  // If the embedder supports it, collect metrics on what modules were resolved.
+  auto metrics = observer.onResolveModule(context.specifier, context.type, context.source);
+
+  switch (context.type) {
+    case ResolveContext::Type::BUNDLE: {
+      // For bundle resolution, we only use Bundle, Builtin, and Fallback bundles,
+      // in that order.
+      KJ_IF_SOME(found, tryFind(context, bundles_[kBundle])) {
+        metrics->found();
+        return found;
+      }
+      KJ_IF_SOME(found, tryFind(context, bundles_[kBuiltin])) {
+        metrics->found();
+        return found;
+      }
+      KJ_IF_SOME(found, tryFind(context, bundles_[kFallback])) {
+        metrics->found();
+        return found;
+      }
+
+      KJ_IF_SOME(parent, maybeParent) {
+        return parent.resolve(context);
+      }
+
+      metrics->notFound();
+      return kj::none;
+     }
+    case ResolveContext::Type::BUILTIN: {
+      // For built-in resolution, we only use builtin and builtin-only bundles.
+      KJ_IF_SOME(found, tryFind(context, bundles_[kBuiltin])) {
+        metrics->found();
+        return found;
+      }
+      KJ_IF_SOME(found, tryFind(context, bundles_[kBuiltinOnly])) {
+        metrics->found();
+        return found;
+      }
+
+      KJ_IF_SOME(parent, maybeParent) {
+        return parent.resolve(context);
+      }
+
+      metrics->notFound();
+      return kj::none;
+    }
+    case ResolveContext::Type::BUILTIN_ONLY: {
+      // For built-in only resolution, we only use builtin-only bundles.
+      KJ_IF_SOME(found, tryFind(context, bundles_[kBuiltinOnly])) {
+        metrics->found();
+        return found;
+      }
+
+      KJ_IF_SOME(parent, maybeParent) {
+        return parent.resolve(context);
+      }
+
+      metrics->notFound();
+      return kj::none;
+    }
+  }
+
+  metrics->notFound();
+  return kj::none;
+}
+
+JsValue ModuleRegistry::resolve(
+    Lock& js,
+    kj::StringPtr specifier,
+    kj::StringPtr exportName,
+    ResolveContext::Type type,
+    ResolveContext::Source source,
+    kj::Maybe<const Url&> maybeReferrer) {
+  auto& bound = IsolateModuleRegistry::from(js.v8Isolate);
+  auto url = ([&]{
```

Comment:
Based on my question about recursive aliasing above: Is that needed here too?

Reply:
Not here. We'll leave it to individual `ModuleBundle`s to deal with aliases.